### DISCUSSION TITLE
error with installation

### DISCUSSION THREAD

**Discussion by Maccchiatooo** [Nek5000/nekRS#552] [Q&A]
I cannot install nekRS for the following errors, appreciate for the help~ Thanks.

n file included from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/utils.hpp:6,
from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/stream.cpp:1:
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:61:23: error: ‘uint32_t’ does not name a type
61 | using return_type = uint32_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:10:1: note: ‘uint32_t’ is defined in header ‘’; did you forget to ‘#include ’?
9 | #include
+++ |+#include
10 | namespace sycl {
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: error: ‘uint64_t’ does not name a type
65 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: error: ‘uint64_t’ does not name a type
69 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:159:41: error: ‘return_type’ in ‘struct sycl::info::device::max_compute_units’ does not name a type
159 | inline info::device::max_compute_units::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:166:39: error: ‘return_type’ in ‘struct sycl::info::device::global_mem_size’ does not name a type
166 | inline info::device::global_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:173:38: error: ‘return_type’ in ‘struct sycl::info::device::local_mem_size’ does not name a type
173 | inline info::device::local_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: error: ‘uint64_t’ does not name a type
276 | uint64_t get_profiling_info() const {
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
In file included from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/utils.hpp:6,
from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/utils.cpp:3:
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:61:23: error: ‘uint32_t’ does not name a type
61 | using return_type = uint32_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:10:1: note: ‘uint32_t’ is defined in header ‘’; did you forget to ‘#include ’?
9 | #include
+++ |+#include
10 | namespace sycl {
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: error: ‘uint64_t’ does not name a type
65 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: error: ‘uint64_t’ does not name a type
69 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
In file included from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/utils.hpp:6,
from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/memoryPool.cpp:1:
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:61:23: error: ‘uint32_t’ does not name a type
61 | using return_type = uint32_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:10:1: note: ‘uint32_t’ is defined in header ‘’; did you forget to ‘#include ’?
9 | #include
+++ |+#include
10 | namespace sycl {
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: error: ‘uint64_t’ does not name a type
65 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: error: ‘uint64_t’ does not name a type
69 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
In file included from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/utils.hpp:6,
from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/buffer.cpp:2:
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:61:23: error: ‘uint32_t’ does not name a type
61 | using return_type = uint32_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:10:1: note: ‘uint32_t’ is defined in header ‘’; did you forget to ‘#include ’?
9 | #include
+++ |+#include
10 | namespace sycl {
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: error: ‘uint64_t’ does not name a type
65 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: error: ‘uint64_t’ does not name a type
69 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:159:41: error: ‘return_type’ in ‘struct sycl::info::device::max_compute_units’ does not name a type
159 | inline info::device::max_compute_units::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:166:39: error: ‘return_type’ in ‘struct sycl::info::device::global_mem_size’ does not name a type
166 | inline info::device::global_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:173:38: error: ‘return_type’ in ‘struct sycl::info::device::local_mem_size’ does not name a type
173 | inline info::device::local_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:159:41: error: ‘return_type’ in ‘struct sycl::info::device::max_compute_units’ does not name a type
159 | inline info::device::max_compute_units::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:166:39: error: ‘return_type’ in ‘struct sycl::info::device::global_mem_size’ does not name a type
166 | inline info::device::global_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:173:38: error: ‘return_type’ in ‘struct sycl::info::device::local_mem_size’ does not name a type
173 | inline info::device::local_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:159:41: error: ‘return_type’ in ‘struct sycl::info::device::max_compute_units’ does not name a type
159 | inline info::device::max_compute_units::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:166:39: error: ‘return_type’ in ‘struct sycl::info::device::global_mem_size’ does not name a type
166 | inline info::device::global_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:173:38: error: ‘return_type’ in ‘struct sycl::info::device::local_mem_size’ does not name a type
173 | inline info::device::local_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: error: ‘uint64_t’ does not name a type
276 | uint64_t get_profiling_info() const {
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: error: ‘uint64_t’ does not name a type
276 | uint64_t get_profiling_info() const {
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: error: ‘uint64_t’ does not name a type
276 | uint64_t get_profiling_info() const {
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
In file included from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/utils.hpp:6,
from /home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/memory.cpp:1:
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:61:23: error: ‘uint32_t’ does not name a type
61 | using return_type = uint32_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:10:1: note: ‘uint32_t’ is defined in header ‘’; did you forget to ‘#include ’?
9 | #include
+++ |+#include
10 | namespace sycl {
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: error: ‘uint64_t’ does not name a type
65 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:65:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: error: ‘uint64_t’ does not name a type
69 | using return_type = uint64_t;
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:69:23: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:159:41: error: ‘return_type’ in ‘struct sycl::info::device::max_compute_units’ does not name a type
159 | inline info::device::max_compute_units::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:166:39: error: ‘return_type’ in ‘struct sycl::info::device::global_mem_size’ does not name a type
166 | inline info::device::global_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:173:38: error: ‘return_type’ in ‘struct sycl::info::device::local_mem_size’ does not name a type
173 | inline info::device::local_mem_size::return_type
| ^~~~~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: error: ‘uint64_t’ does not name a type
276 | uint64_t get_profiling_info() const {
| ^~~~~~~~
/home/chunheng/nekRS/3rd_party/occa/src/occa/internal/modes/dpcpp/polyfill.hpp:276:3: note: ‘uint64_t’ is defined in header ‘’; did you forget to ‘#include ’?
gmake[2]: *** [3rd_party/occa/CMakeFiles/libocca.dir/build.make:2946: 3rd_party/occa/CMakeFiles/libocca.dir/src/occa/internal/modes/dpcpp/stream.cpp.o] Error 1
gmake[2]: *** Waiting for unfinished jobs....
gmake[2]: *** [3rd_party/occa/CMakeFiles/libocca.dir/build.make:2862: 3rd_party/occa/CMakeFiles/libocca.dir/src/occa/internal/modes/dpcpp/buffer.cpp.o] Error 1
gmake[2]: *** [3rd_party/occa/CMakeFiles/libocca.dir/build.make:2974: 3rd_party/occa/CMakeFiles/libocca.dir/src/occa/internal/modes/dpcpp/utils.cpp.o] Error 1
gmake[2]: *** [3rd_party/occa/CMakeFiles/libocca.dir/build.make:2918: 3rd_party/occa/CMakeFiles/libocca.dir/src/occa/internal/modes/dpcpp/memoryPool.cpp.o] Error 1
gmake[2]: *** [3rd_party/occa/CMakeFiles/libocca.dir/build.make:2904: 3rd_party/occa/CMakeFiles/libocca.dir/src/occa/internal/modes/dpcpp/memory.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/Makefile2:530: 3rd_party/occa/CMakeFiles/libocca.dir/all] Error 2
gmake: *** [Makefile:156: all] Error 2
chunheng@chunheng-XPS:~/nekRS$

**Comment by joneuhauser** [comment]
Add "#include <cstdint>" in line 10 of polyfill.hpp, or use the next branch.

**Comment by yslan** [comment]
What nekrs version / machine / backend / compilers are you using?

If it's v23 (repo/master) + HIP + Frontier, as @joneuhauser pointed out, the error will probably fixed by a newer OCCA. You can either try repo/next branch or adapt this patch:
https://github.com/yslan/nekRS/commit/78504dd3852089f703012a2285dd5e7ab360b01d

**Comment by VictorEijkhout** [comment]
@Maccchiatooo run this on the unpacked source:
```
find . -name polyfill.hpp -exec sed -i -e 's/ uint/ std::uint/g' {} \;
```
@yslan the last release is two years old. Time for a new release with this issue fixed?

**Reply by yslan** [reply]
We are currently working on a major release, v25. The beta version might be pushed to the repo in about 1-2 weeks. 
Before v25 is available and well-tested, I'd recommend `v24-dev`, which has the updated occa version addressing this issue (along with other known fixes).
https://github.com/Nek5000/nekRS/tree/v24-development

**Comment by VictorEijkhout** [comment]
Good to hear that a new release is coming up. If you make an actual rc release I will be notified about it.

Btw, I was able to build the "next" branch of the repo.
